src/wallet/rpcwallet.cpp

```diff
@@ -3520,6 +3520,336 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        pwallet->GetKey(meta.hdMasterKeyID, key);
+        CExtKey masterKey;
+        masterKey.SetMaster(key.begin(), key.size());
+        // Add to map
+        keypath.insert(keypath.begin(), masterKey.key.GetPubKey().GetID().GetUint32(0));
+        hd_keypaths.emplace(vchPubKey, keypath);
+    }
+}
+
+void fill_psbt(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const CTransaction* txConst)
+{
+    // Get all of the previous transactions
+    bool psbtx_blank = psbtx.IsNull();
+    for (unsigned int i = 0; i < txConst->vin.size(); ++i) {
+        CTxIn txin = txConst->vin[i];
+        PartiallySignedInput input;
+        if (!psbtx_blank) {
+            input = psbtx.inputs.at(i);
+        }
+
+        // If this input is not empty, skip it
+        if (!psbtx_blank && (!input.IsNull() || txin.scriptSig.empty() || txin.scriptWitness.IsNull())) {
+            continue;
+        }
+
+        uint256 txhash = txin.prevout.hash;
+
+        // If we don't know about this input, skip it and let someone else deal with it
+        if (!pwallet->mapWallet.count(txhash)) {
+            if (psbtx_blank) {
+                psbtx.inputs.push_back(input);
+            }
+            continue;
+        }
+        const CWalletTx& wtx = pwallet->mapWallet.at(txhash);
+        const CTransaction& ctx = *wtx.tx;
+
+        // Get scriptpubkey and check for redeemScript or witnessscript
+        CTxOut prevout = ctx.vout[txin.prevout.n];
+        txnouttype type;
+        std::vector<std::vector<unsigned char>> solns;
+        Solver(prevout.scriptPubKey, type, solns);
+        // Get script hashes
+        if (type == TX_SCRIPTHASH) {
+            // get the hash and find it in the wallet.
+            CScript redeem_script;
+            uint160 hash(solns[0]);
+            pwallet->GetCScript(CScriptID(hash), redeem_script);
+
+            // put redeem_script in map
+            psbtx.redeem_scripts.emplace(hash, redeem_script);
+
+            // Now check whether the redeem_script is a witness script
+            solns.clear();
+            Solver(redeem_script, type, solns);
+        }
+        // Get witness scripts
+        bool witness = false;
+        if (type == TX_WITNESS_V0_SCRIPTHASH) {
+            witness = true;
+            // Get the hash from the solver return
+            uint160 hash;
+            CRIPEMD160().Write(&solns[0][0], solns[0].size()).Finalize(hash.begin());
+
+            // Lookup hash from wallet
+            CScript witness_script;
+            pwallet->GetCScript(CScriptID(hash), witness_script);
+
+            // Put witness script in map
+            uint256 hash256(solns[0]);
+            psbtx.witness_scripts.emplace(hash256, witness_script);
+
+            // Decode the witness script
+            solns.clear();
+            Solver(witness_script, type, solns);
+        }
+        // Get public keys if hd is enabled
+        if (pwallet->IsHDEnabled()) {
+            if (type == TX_PUBKEYHASH) {
```

Comment:
should be:

`if (type == TX_PUBKEYHASH || type == TX_WITNESS_V0_KEYHASH) {`

Reply:
Done.